Node E2E: Prep for continuous Docker validation node e2e test

Random-Liu opened the pull request:
Based on https://github.com/kubernetes/kubernetes/pull/28516, for https://github.com/kubernetes/kubernetes/issues/25215.

https://github.com/kubernetes/kubernetes/pull/26813 added support to run e2e test on gci preview image and newest docker version.
This PR added the same support to node e2e test.

The main dependencies of node e2e test are `docker`, `kubelet`, `etcd` and `apiserver`.
Currently, node e2e test builds `kubelet` and `apiserver` locally, and copies them into `/tmp` directory in VM instance. GCI also has built-in `docker`. So the only dependency missing is `etcd`.

This PR injected a simple cloud-init script when creating instance to install `etcd` during node startup.

@andyzheng0831 for the cloud init script.
@wonderfly for the gci instance setup.
@pwittrock for the node e2e test change.

/cc @dchen1107 

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

Random-Liu committed:
Add docker validation node e2e test.

wonderfly commented on test/e2e_node/jenkins/jenkins-docker-validation.properties:
```diff
@@ -0,0 +1,15 @@
+GCI_IMAGE_PROJECT=container-vm-image-staging
+GCI_IMAGE_FAMILY=gci-preview-test
+GCI_IMAGE=$(gcloud compute images describe-from-family ${GCI_IMAGE_FAMILY} --project=${GCI_IMAGE_PROJECT} --format="value(name)")
+DOCKER_VERSION=$(curl -fsSL --retry 3 https://api.github.com/repos/docker/docker/releases | tac | tac | grep -m 1 "\"tag_name\"\:" | grep -Eo "[0-9\.rc-]+")
```
If the Jenkins node that you use has `jq` installed, this parsing can be largely simplified.

Random-Liu replied:
@wonderfly Agree. We could download it ourselves in this script, but that will introduce another dependency and make the code a little more complex. :)

vishh replied:
If we are using `jq` we should download it as part of this test suite.

ixdy replied:
The `kubekins-test` image has `jq` preinstalled. I've been meaning to get the node e2e test running inside the `kubekins-test` image, but that hasn't happened yet.

Random-Liu replied:
@ixdy one of the purposes of node e2e is just to test different images with different os distros. :)

wonderfly commented:
LGTM for the GCI related settings. Can't comment on others.

ixdy commented:
@k8s-bot test this please, issue #IGNORE

(Sorry, temporarily broke Jenkins.)

Random-Liu commented:
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/27498

vishh commented on test/e2e_node/jenkins/jenkins-docker-validation.properties:
```diff
@@ -0,0 +1,15 @@
+GCE_IMAGE_PROJECT=${GCI_IMAGE_PROJECT}
+GCE_ZONE=us-central1-f
+GCE_PROJECT=kubernetes-jenkins
+GCE_INSTANCE_METADATA="user-data<${GCI_CLOUD_INIT},gci-ensure-gke-docker=true,gci-docker-version=${DOCKER_VERSION}"
```
nit: Can you add some description for these flags?

Random-Liu replied:
ACK.

vishh commented:
Just one nit. Otherwise LGTM

Random-Liu commented:
@vishh Addressed comments and will squash before merging. PTAL~

BTW, the commit `Disable kubenet for local node e2e test.` is from #28516. We need it in this PR because the file system of GCI is read only, we can not download CNI to `/opt/bin`. #28516 moves CNI to `/tmp` directory which solves the problem.

vishh commented on hack/e2e-node-test.sh:
```diff
@@ -129,6 +129,6 @@ else
   # Provided for backwards compatibility
   "${ginkgo}" --focus=$focus --skip=$skip "${KUBE_ROOT}/test/e2e_node/" --report-dir=${report} \
-    -- --alsologtostderr --v 2 --node-name $(hostname) --build-services=true --start-services=true --stop-services=true
+    -- --alsologtostderr --v 2 --node-name $(hostname) --disable-kubenet=true --build-services=true --start-services=true --stop-services=true
```
Why disable kubenet? Why not download it instead?

Random-Liu replied:
This is just for local node e2e test #28516 -> `make test_e2e_node`.
For local development, it is a little complicated to download cni and nsenter, because the developer may have different environment. At least the nsenter installation script doesn't work on my desktop. :(

vishh replied:
I'm concerned that a bug in networking will not show up with local testing. As much as possible, we want the local and remote testing environment to be the same. WDYT?

vishh replied:
We already require etcd to be installed. Can we also require cni plugin? or may be auto-download it?

Random-Liu commented:
@vishh I've rebased the PR, the disable kubenet locally commit is never in this PR. Can we get this merged first? And I'll send another PR to document the kubenet thing. :)
